### PR TITLE
Implement NoScan and add BaseScan

### LINKED ISSUE #256: Add NoScan Entry at builder gui
### Feature you want to implement
First, add a NoScan class. 
Before, the builder GUI only supported RangeScan. When NoScan class is added, feature for selecting the type of scan we want is needed. So this feature will also be added.
Since Scan classes share common method and some variables, it seems that base class for all Scan classes is needed.

### How the feature is implemented
1. NoScan has two important parameters, `value` and `repetitions`. Use `QDoubleSpinBox` for control `value`, and `QSpinBox` for `repetitions`. It has similar structure with `RangeScan` class.
2. Since we added NoScan class, ScanEntry class should be updated. Use `QStackedWidget` to store each scan type entry. And Use `QRadioButton` to make selection button gui. All radio buttons are added at `QHBoxLayout`. If one of button is checked by user, method `scanTypeToggled` is called and update current widget at ScanEntry. Original iquip code used `QButtonGroup`, but as I checked there was no reason to use it.
3. Add BaseScan class that is the base class for Scan classes. Since all Scan classes use `applyProperties` method to initialize there widget, BaseScan class can have that method.

### PR DESCRIPTION
**Pull request by Aijuh**
This closes #256
##
1. `_NoScan` class has been added. It has two arguments, `value` and `repetitions`. 
2. Implment scan type selection in the `_ScanEntry` class. You can choose RangeScan or NoScan at our iquip GUI. Two widgets are contained at `stackWidget`. `radioButtons` is used to select between two scan type.
3. Added BaseScan that is base class for all scan widgets. Since some attributes and method are duplicated, it seems more efficient way.
##
Below is the image of current figure of builder GUI.
image

### CODE REVIEW

**Comment by BECATRUE** [comment]
Can you indicate the relative issue in PR contents, like "This is related to #{issue}."?

**Comment by BECATRUE** [comment]
Then, you should write how to implement it in as possible as detailed in the issue, before implementing it.
From the next PR, please keep this in your mind.

File: iquip/apps/builder.py [hunk 1]
```diff
@@ -322,11 +323,27 @@ def __init__(self, name: str, argInfo: Dict[str, Any], parent: Optional[QWidget]
         super().__init__(name, argInfo, parent=parent)
         self.state = self.get_state()
         procdesc = self.get_procdesc()
-        # TODO(AIJUH): Add features for stack widget.
         self.stackWidget = QStackedWidget(self)
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: This is just a suggestion.
As the class name is `QStackedWidget`, how about renaming it to `stackedWidget`?

If you prefer to the original one, don't keep in your mind.

**Comment by BECATRUE** [comment]
I'm sorry that I couldn't check that methods and variables related to PyQt should have names with camel case, due to PyQt customary practice.

Please rename them implemented in the previous and this PRs.

File: iquip/apps/builder.py [hunk 2]
```diff
@@ -322,11 +323,27 @@ def __init__(self, name: str, argInfo: Dict[str, Any], parent: Optional[QWidget]
...
+            self.stackWidget.addWidget(widget)
+        self.radioButtons = OrderedDict()
+        self.radioButtons["NoScan"] = QRadioButton("No scan")
+        self.radioButtons["RangeScan"] = QRadioButton("Range")
+        scan_type = QButtonGroup(self)
+        buttonLayout = QHBoxLayout()
+        for _, b in enumerate(self.radioButtons.values()):
+            buttonLayout.addWidget(b)
+            scan_type.addButton(b)
+            b.toggled.connect(self.scan_type_toggled)
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: Umm.. I think you can write this code in only one loop.

For example,
```python
self.widgets = {}
scanButtonGroup = QButtonGroup(self)
buttonLayout = QHBoxLayout()
for ty, (scanCls, buttonName) in {"NoScan": (_NoScan, "No scan"), ...}.items():
  scanWidget = scanCls(procdesc, self.state[ty])
  self.widgets[ty] = scanWidget
  self.stackWidget.addWidget(scanWidget)
  button = QRadioButton(buttonName)
  scanButtonGroup.addButton(button, id)  # id is defined in IntEnum class.
scanButtonGroup.buttonToggled.connect(...)  # this is called whenever any button is toggled.
```

Comment (BECATRUE) [reply iquip/apps/builder.py]: I think it has some advantages:
1. `OrderedDict` is unnecessary.
2. Easy to manage each widget.
3. Avoid to cause an error due to a typo.

Comment (BECATRUE) [reply iquip/apps/builder.py]: And please check the followings:
1. The name of `widgets` is too vague. How about "scanWidgets"?
2. If you use an id and enum class to identify which button is toggled, it seems that `radioButtons` is unnecessary.

**Comment by Aijuh** [comment]
I understood. I will make new issue and convert to camel case.

**Comment by Aijuh** [comment]
@BECATRUE  I updated it.

**Comment by BECATRUE** [comment]
> I understood. I will make new issue and convert to camel case.

You already fixed it and won't make a new issue, right?

File: iquip/apps/builder.py [hunk 3]
```diff
@@ -320,16 +321,30 @@ def __init__(self, name: str, argInfo: Dict[str, Any], parent: Optional[QWidget]
...
+        self.state = self.getState()
+        procdesc = self.getProcdesc()
         self.stackWidget = QStackedWidget(self)
-        self.rangeWidget = _RangeScan(procdesc, self.state["RangeScan"])
+        self.widgets = OrderedDict()
+        self.widgets["NoScan"] = _NoScan(procdesc, self.state["NoScan"])
+        self.widgets["RangeScan"] = _RangeScan(procdesc, self.state["RangeScan"])
+        for widget in self.widgets.values():
+            self.stackWidget.addWidget(widget)
+        self.radioButtons = OrderedDict()
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: Why did you remove the button group?
For radio buttons to operate in the same group, isn't it necessary?

File: iquip/apps/builder.py [hunk 4]
```diff
@@ -320,16 +321,30 @@ def __init__(self, name: str, argInfo: Dict[str, Any], parent: Optional[QWidget]
...
+        self.radioButtons["NoScan"] = QRadioButton("No scan")
+        self.radioButtons["RangeScan"] = QRadioButton("Range")
+        buttonLayout = QHBoxLayout()
+        for _, b in enumerate(self.radioButtons.values()):
+            buttonLayout.addWidget(b)
+            b.toggled.connect(self.scanTypeToggled)
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: This code can be simplified.

For example,
```python
# in this class
class ScanType(IntEnum):
  NoScan = 0
  RangeScan = 1

# in this method
self.widgets = {}
buttonGroup = QButtonGroup(self)
buttonLayout = QHBoxLayout()
for ty, (scanCls, buttonName) in {"NoScan": (_NoScan, "No scan"), ...}.items()
  scanWidget = scanCls(procdesc, self.state[ty])
  self.widgets[ty] = scanWidget
  self.stackWidget.addWidget(scanWidget)
  button = QRadioButton(buttonName)
  buttonGroup.addButton(button, _ScanEntry.ScanType[ty].value)
  buttonLayout.addWidget(button)
buttonGroup.buttonClicked.connect(...)
```

I think it has some advantages:
1. Shorten the code
2. Easy to change the order of the scan widgets.
3. Make the management easy and avoid to cause an error due to typos.
4. Use only one signal, i.e. `buttonGroup.buttonClicked`.

Of course, the code may have some errors, hence please recheck it.

Comment (BECATRUE) [reply iquip/apps/builder.py]: And, please check the followings:
1. The name of `self.widgets` looks too vague. How about `self.scanWidgets`?
2. Is `self.radioButtons` necessary?

File: iquip/apps/builder.py [hunk 5]
```diff
@@ -375,17 +390,119 @@ def value(self) -> Dict[str, Any]:
...
+            if button.isChecked():
+                self.stackWidget.setCurrentWidget(self.widgets[ty])
+                self.state["selected"] = ty
+                break
+
 
-class _RangeScan(QWidget):
-    """A widget for range scan in _ScanEntry.
+class _BaseScan(QWidget):
+    """Base class for all scannable widget.
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: "a scannable widget" or "all scannble widgets".

File: iquip/apps/builder.py [hunk 6]
```diff
@@ -375,17 +390,119 @@ def value(self) -> Dict[str, Any]:
...
+        """
+        super().__init__(procdesc, parent=parent)
+        self.valueSpinBox = QDoubleSpinBox(self)
+        self.applyProperties(self.valueSpinBox, procdesc)
+        self.valueSpinBox.setValue(state["value"]/self.scale)
+        repetitionsSpinBox = QSpinBox(self)
+        repetitionsSpinBox.setMinimum(1)
+        repetitionsSpinBox.setMaximum((1 << 31) - 1)
+        repetitionsSpinBox.setValue(state["repetitions"])
+        #layout
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: "# layout"

File: iquip/apps/builder.py [hunk 7]
```diff
@@ -375,17 +390,119 @@ def value(self) -> Dict[str, Any]:
...
+        Args:
+            procdesc: Each key and its value are:
+              unit, scale, global_step, global_min, global_max, ndecimals: 
+                See argInfo in _ScanEntry.__init__().
+        """
+        super().__init__(parent=parent)
+        self.scale = procdesc["scale"]
+        self.layout = QGridLayout(self)
+
+    def applyProperties(self, widget: QDoubleSpinBox, procdesc: Dict[str, Any]):
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: You moved this method to the super class, so it looks better to make `procdesc` an attribute.
Then, you don't need to call this method with it.

File: iquip/apps/builder.py [hunk 8]
```diff
@@ -375,17 +390,119 @@ def value(self) -> Dict[str, Any]:
...
+        if minVal is None:
+            minVal = 0.0
+        if maxVal is None:
+            maxVal = 99.99
+        widget.setMinimum(minVal / self.scale)
+        widget.setMaximum(maxVal / self.scale)
+        widget.setSuffix(unit)
+        widget.setSingleStep(step / self.scale)
+
+    def getScanArgs(self) -> Dict[str, Any]:
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: I remembered that Jiyong said not to start a method name with "get", according to the PyQt rule.
So, can you rename these three methods, `getState()`, `getProcdesc()`, and `getScanArgs`.

For example, use `scanArguments()` instead of it.
(I think `scanArguments` is better that `scanArgs` because we have used it as the full name.)

However, I'm worried to rename `getState()` to `state()` because it already exists as an attribute.
Hence you need to come up with a better idea.

File: iquip/apps/builder.py [hunk 9]
```diff
@@ -375,17 +390,119 @@ def value(self) -> Dict[str, Any]:
...
+        """Returns the arguments of the scannable object.
+        
+        This must be overridden in the subclass.
+        """
+        raise NotImplementedError
+
+
+class _NoScan(_BaseScan):
+    """An widget for no scan in _ScanEntry.
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: "A widget". But, I think "a" is unnecessary like other classes.

File: iquip/apps/builder.py [hunk 10]
```diff
@@ -396,27 +513,23 @@ def __init__(
         """Extended.
 
         Args:
-            procdesc: Each key and its value are:
-              unit, scale, global_step, global_min, global_max, ndecimals: 
-                See argInfo in _ScanEntry.__init__().
+            procdesc: See the __init__() in _BaseScan.
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: As I think, any explanation about `procdesc` is unnecessary becuase "Extended." indicates that there are more explanations in `_BaseScan.__init__()`.

File: iquip/apps/builder.py [hunk 11]
```diff
@@ -375,17 +390,119 @@ def value(self) -> Dict[str, Any]:
...
+        self,
+        procdesc: Dict[str, Any],
+        state: Dict[str, Any],
+        parent: Optional[QWidget] = None
+        ):
+        """Extended.
+
+        Args:
+            procdesc: See the __init__() in _BaseScan.
+            state:
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: Please add "Each key and its value are:".
And the name of keys looks wrong.

File: iquip/apps/builder.py [hunk 12]
```diff
@@ -396,27 +513,23 @@ def __init__(
...
         Args:
-            procdesc: Each key and its value are:
-              unit, scale, global_step, global_min, global_max, ndecimals: 
-                See argInfo in _ScanEntry.__init__().
+            procdesc: See the __init__() in _BaseScan.
             state: Each key and its value are:
               start: The start point for the RangeScan sequence.
               stop: The end point for the RangeScan sequence.
               npoints: The number of points in the RangeScan sequence.
               randomize: A boolean value that decides whether to shuffle the RangeScan sequence.
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: Can you start with "The" like other keys.

File: iquip/apps/builder.py [hunk 13]
```diff
@@ -375,17 +390,119 @@ def value(self) -> Dict[str, Any]:
...
+        state: Dict[str, Any],
+        parent: Optional[QWidget] = None
+        ):
+        """Extended.
+
+        Args:
+            procdesc: See the __init__() in _BaseScan.
+            state:
+              valueSpinBox: The repeated value in the NoScan sequence.
+              repetitionsSpinBox: A number to repeat the value in the NoScan sequence.
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: Can you start with "The" like the other.

File: iquip/apps/builder.py [hunk 14]
```diff
@@ -375,17 +390,119 @@ def value(self) -> Dict[str, Any]:
...
+        """
+        super().__init__(procdesc, parent=parent)
+        self.valueSpinBox = QDoubleSpinBox(self)
+        self.applyProperties(self.valueSpinBox, procdesc)
+        self.valueSpinBox.setValue(state["value"]/self.scale)
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: Please add a space before and after the operator.

File: iquip/apps/builder.py [hunk 15]
```diff
@@ -375,17 +390,119 @@ def value(self) -> Dict[str, Any]:
...
+        repetitionsSpinBox = QSpinBox(self)
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: Did you intend "self.repetitionsSpinBox"?

File: iquip/apps/builder.py [hunk 16]
```diff
@@ -375,17 +390,119 @@ def value(self) -> Dict[str, Any]:
...
+        self.layout.addWidget(repetitionsSpinBox, 1, 1)
+
+    def getScanArgs(self) -> Dict[str, Any]:
+        """Overridden.
+        
+        Returns the arguments of the no scan.
+        """
+        return {
+            "value": self.valueSpinBox.value(),
+            "repetitions": self.repetitionsSpinBox.value()
```
Comment (BECATRUE) [inline iquip/apps/builder.py:512]: Does it work well..? I think there is no `self.repetitionsSpinBox`.

**Review by BECATRUE — CHANGES_REQUESTED** [review]

File: iquip/apps/builder.py (hunk 8 above)
Comment (Aijuh) [reply iquip/apps/builder.py]: I remembered that you send me the link for rules for coding pyQt. But I cannot find it.
Can you send the link if you have it?
If not, you don't have to. Of course I will fix this case.

File: iquip/apps/builder.py (hunk 3 above)
Comment (Aijuh) [reply iquip/apps/builder.py]: In previous code it works well without button group. Since we operate on QRadioButton seperately.

File: iquip/apps/builder.py (hunk 4 above)
Comment (Aijuh) [reply iquip/apps/builder.py]: I will think in this way.

File: iquip/apps/builder.py (hunk 16 above)
Comment (Aijuh) [reply iquip/apps/builder.py:512]: After I finished testing, I found that the name of Qspinbox for `repetitions` is just repetitions.
There was mistake while I updating it.

File: iquip/apps/builder.py (hunk 15 above)
Comment (Aijuh) [reply iquip/apps/builder.py]: Yes...

File: iquip/apps/builder.py (hunk 14 above)
Comment (Aijuh) [reply iquip/apps/builder.py]: I didn't find it.

File: iquip/apps/builder.py (hunk 10 above)
Comment (Aijuh) [reply iquip/apps/builder.py]: I see.

File: iquip/apps/builder.py (hunk 4 above)
Comment (Aijuh) [reply iquip/apps/builder.py]: Thank for giving entire code.

File: iquip/apps/builder.py (hunk 3 above)
Comment (BECATRUE) [reply iquip/apps/builder.py]: It looks natural that they are tied by a button group.
However, you already fixed to use it, thus please never mind.

File: iquip/apps/builder.py [hunk 17]
```diff
@@ -301,10 +301,14 @@ class _ScanEntry(_BaseEntry):
...
-        rangeWidget: TODO(AIJUH): The Widget that will be removed at next PR.
+        stackedWidget: The QStackedWidget that contains widgets of each scannable type.
+        scanButtonGroup: The QButtonGroup that groups QRadiobuttons for selecting scan widget.
+        scanWidgetsDict: The dictionary that contains each scan widget.
+        scanButtonsDict: The dictionary that describes id of QRadiobutton at scanButtonGroup.
     """
-    def __init__(self, name: str, argInfo: Dict[str, Any], parent: Optional[QWidget] = None):
+    def __init__(
+        self, name: str, argInfo: Dict[str, Any], parent: Optional[QWidget] = None
+        ): # pylint: disable=too-many-locals
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: I think this Pylint comment can be moved to the definition above.
And the definition can be written in a line.

File: iquip/apps/builder.py (hunk 4 above)
Comment (BECATRUE) [reply iquip/apps/builder.py]: The reason I suggested to use an enum class is that it looks inefficient to check all buttons in `scanTypeToggled()`.
According to [QButtonGroup](https://doc.qt.io/qtforpython-5/PySide2/QtWidgets/QButtonGroup.html#detailed-description), the `buttonClicked` signal is emitted with id of the clicked button, so you can make code more efficiently.

For example,
```python
def scanButtonClicked(self, id: int):
  ty = _ScanEntry.ScanType(id).name
  stackedWidget = self.scanWidgetsDict[ty]
  ...
```

File: iquip/apps/builder.py [hunk 18]
```diff
@@ -320,16 +324,33 @@ def __init__(self, name: str, argInfo: Dict[str, Any], parent: Optional[QWidget]
...
-        # TODO(AIJUH): Add features for stack widget.
-        self.stackWidget = QStackedWidget(self)
-        self.rangeWidget = _RangeScan(procdesc, self.state["RangeScan"])
+        self.state = self.stateManager()
+        procdesc = self.procdescManager()
+        self.stackedWidget = QStackedWidget(self)
+        buttonLayout = QHBoxLayout()
+        self.scanButtonGroup = QButtonGroup(self)
+        scanDict = {"NoScan": (_NoScan, "No scan"), "RangeScan": (_RangeScan, "Range")}
+        self.scanWidgetsDict = {}
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: It looks sufficient to name it to `scanWidgets`.
In other apps, we have hardly indicated "Dict".

File: iquip/apps/builder.py [hunk 19]
```diff
@@ -320,16 +324,33 @@ def __init__(self, name: str, argInfo: Dict[str, Any], parent: Optional[QWidget]
...
+        selectedButton.setChecked(True)
+        # layout
         scanLayout = QVBoxLayout()
-        scanLayout.addWidget(self.rangeWidget)
+        scanLayout.addLayout(buttonLayout)
+        scanLayout.addWidget(self.stackedWidget)
         self.layout.addLayout(scanLayout)
 
-    def get_state(self) -> Dict[str, Any]:
+    def stateManager(self) -> Dict[str, Any]:
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: First, we have usually named a method to start with a verb, except to a slot.
Also `*Manager` feels like it means a manager object.

Then, how about just `initState()` or something better?

File: iquip/apps/builder.py [hunk 20]
```diff
@@ -375,17 +396,120 @@ def value(self) -> Dict[str, Any]:
...
+        layout: The outermost layout.
+        procdesc: Each key and its value are:
+          unit, scale, global_step, global_min, global_max, ndecimals: 
+            See argInfo in _ScanEntry.__init__().
+    """
+    def __init__(self, procdesc: Dict[str, Any], parent: Optional[QWidget] = None):
+        """Extended.
+
+        Args:
+            procdesc: See procdesc in __init__().
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: "procdesc: See the attributes section." or
"See the attributes section."

File: iquip/apps/builder.py (hunk 8 above)
Comment (BECATRUE) [reply iquip/apps/builder.py]: Umm.. I don't remember that I sent you about this😅
About Python coding style, you can refer to this [styleguide](https://google.github.io/styleguide/pyguide.html), but I don't know about PyQt coding style..

File: iquip/apps/builder.py [hunk 21]
```diff
@@ -375,17 +396,120 @@ def value(self) -> Dict[str, Any]:
...
+        Returns the arguments of the no scan.
+        """
+        return {
+            "value": self.valueSpinBox.value(),
+            "repetitions": self.repetitionsSpinBox.value()
+        }
+
+
+class _RangeScan(_BaseScan):
+    """A widget for range scan in _ScanEntry.
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: Please don't write "A" like `_NoScan`.

File: iquip/apps/builder.py (hunk 15 above)
Comment (BECATRUE) [reply iquip/apps/builder.py]: Did you apply this review?

**Comment by Aijuh** [comment]
Yes I will apply your current review.

File: iquip/apps/builder.py (hunk 19 above)
Comment (Aijuh) [reply iquip/apps/builder.py]: I also thought abot that.
I thought starting with `init` might cause problem.

**Comment by Aijuh** [comment]
I fixed codes. In terms of simplifying signal, I used `QButtonGroup.checkedId()` method.

**Comment by BECATRUE** [comment]
> I fixed codes. In terms of simplifying signal, I used QButtonGroup.checkedId() method.

Still, I think using an `Enum` class is better because we can check each mapping directly.

Also, you should consider which one between `buttonClicked` and `buttonToggled` signal is better.
The former is emitted with only clicked id (or button), and the latter is emitted with clicked id (or button) and its status.
In other words, you DON'T need to check what button is clicked in `scanTypeToggled` and it is okay to just check its argument.

File: iquip/apps/builder.py (hunk 15 above)
Comment (BECATRUE) [reply iquip/apps/builder.py]: @Aijuh

**Comment by Aijuh** [comment]
Updated code to use `enum` (`idToButton`}, that mapps id of button to selected widget's name.
But There was already a dictionary{`scanButtonsDict`} that mapps widget's name to id. Two dictionary have exact information. I want to get rid of one of them. But in order to get key from value, iteration is also needed.
If we remove `scanButtonsDict`, iteration occur only once at the `__init__` method. So it seems more preferable.

And buttonClicked signal emits `QRadioButton` that is clicked, not `id` itself. Even if it can be changed it to emit `int` by `buttonClicked[int].connect`, it caused pylint error.

File: iquip/apps/builder.py [hunk 22]
```diff
@@ -375,17 +398,119 @@ def value(self) -> Dict[str, Any]:
         selected = self.state["selected"]
         return self.state[selected]
 
+    def scanTypeClicked(self, selectedButton: QRadioButton):
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: Can you check if a `pyqtSlot()` decorator is available?

**Comment by BECATRUE** [comment]
I also think removing the former is efficient if you can remove `scanButtonsDict` or an enum class.

If you use an enum, you can simplify the loop in `__init__()` as follows:
```python
class ScanType(IntEnum):
  NoScan = 0
  RangeScan = 1

def __init__(self, ...):
  ...
  for scanType in _ScanEntry.ScanType:
    ty = scanType.name
    buttonId = scanType.value
    scanCls, buttonName = scanDict[ty]
    ...
```

It makes one easy to change the order of the scan widgets and buttons.

**Review by BECATRUE — CHANGES_REQUESTED** [review]

**Comment by Aijuh** [comment]
I didn't have experience with enum in python and misunderstand the concept.
I thought that `Enum` is only used at C language.
I will upgrade code by using `Enum`.

**Comment by Aijuh** [comment]
I modified the code to use `Enum` class. It simplifies duplicated data. 

In terms of `pyqtSlot`, I should have imported `QAbstractButton` class to add it.

When I defined key of `Enum` class as `NoScan` and `RangeScan`, pylint told that camel case is bad convention. So I disabled pylint for now.

File: iquip/apps/builder.py [hunk 23]
```diff
@@ -3,13 +3,14 @@
 import json
 import logging
 from typing import Any, Dict, Optional, Tuple, Union
+from enum import IntEnum, unique
```
Comment (BECATRUE) [inline iquip/apps/builder.py:5]: Please move this before `from typing ...` for alphabetical order.

File: iquip/apps/builder.py [hunk 24]
```diff
@@ -301,9 +302,17 @@ class _ScanEntry(_BaseEntry):
           "selected": The name of the selected scannable type.
           "NoScan", "RangeScan", "CenterScan", and "ExplicitScan": The dictionary that contains 
             argument info of the corresponding scannable type.
-        stackWidget: The QStackedWidget that contains widgets of each scannable type.
-        rangeWidget: TODO(AIJUH): The Widget that will be removed at next PR.
+        stackedWidget: The QStackedWidget that contains widgets of each scannable type.
+        scanButtonGroup: The QButtonGroup that groups QRadiobuttons for selecting scan widget.
+        scanWidgets: The dictionary that contains each scan widget.
     """
+    @unique
```
Comment (BECATRUE) [inline iquip/apps/builder.py:310]: 👍

File: iquip/apps/builder.py [hunk 25]
```diff
@@ -301,9 +302,17 @@ class _ScanEntry(_BaseEntry):
...
+    # pylint: disable=invalid-name
+    class ScanType(IntEnum):
```
Comment (BECATRUE) [inline iquip/apps/builder.py:312]: Please add an empty line before and after the class definition.

File: iquip/apps/builder.py [hunk 26]
```diff
@@ -355,7 +384,7 @@ def get_state(self) -> Dict[str, Any]:
                     state[ty] = default
```
Comment (BECATRUE) [inline iquip/apps/builder.py:386]: Can you replace the list `["NoScan", "RangeScan", "CenterScan", "ExplicitScan"]` with a tuple, in the line above?

File: iquip/apps/builder.py [hunk 27]
```diff
@@ -320,16 +329,36 @@ def __init__(self, name: str, argInfo: Dict[str, Any], parent: Optional[QWidget]
...
+        scanDict = {"NoScan": (_NoScan, "No scan"), "RangeScan": (_RangeScan, "Range")}
+        self.scanWidgets = {}
+        for scanType in _ScanEntry.ScanType:
+            ty = scanType.name
+            buttonId = scanType.value
+            scanCls, buttonName = scanDict[ty]
+            scanWidget = scanCls(procdesc, self.state[ty])
+            self.stackedWidget.addWidget(scanWidget)
+            self.scanWidgets[ty] = scanWidget
+            button = QRadioButton(buttonName)
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: Is there any reason not to set the parent to `self`?

File: iquip/apps/builder.py [hunk 28]
```diff
@@ -320,16 +329,36 @@ def __init__(self, name: str, argInfo: Dict[str, Any], parent: Optional[QWidget]
...
+            buttonLayout.addWidget(button)
+            self.scanButtonGroup.addButton(button, buttonId)
+        self.scanButtonGroup.buttonClicked.connect(self.scanTypeClicked)
+        selected = self.state["selected"]
+        self.stackedWidget.setCurrentWidget(self.scanWidgets[selected])
+        selectedEnum = _ScanEntry.ScanType[selected]
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: `selectedScanType` or `selectedType` looks better.

File: iquip/apps/builder.py [hunk 29]
```diff
@@ -375,17 +404,120 @@ def value(self) -> Dict[str, Any]:
         selected = self.state["selected"]
         return self.state[selected]
 
+    @pyqtSlot(QAbstractButton)
+    def scanTypeClicked(self, selectedButton: QAbstractButton):
+        """Switches current scan widget at stacked layout in _ScanEntry.
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: How about "in the stacked layout."?

File: iquip/apps/builder.py [hunk 30]
```diff
@@ -375,17 +404,120 @@ def value(self) -> Dict[str, Any]:
...
+        
+        Once the checked button at radiobutton group clicked, this is called.
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: It looks a little weird.
How about "Once a scan type button is clicked, ..."?

File: iquip/apps/builder.py [hunk 31]
```diff
@@ -375,17 +404,120 @@ def value(self) -> Dict[str, Any]:
...
+
+        Attributes:
+            selectedButton: A QRadioButton that is clicked.
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: It can be simplified to "The clicked QRadioButton.".

File: iquip/apps/builder.py [hunk 32]
```diff
@@ -375,17 +404,120 @@ def value(self) -> Dict[str, Any]:
...
+        """
+        selectedEnum = _ScanEntry.ScanType(self.scanButtonGroup.id(selectedButton))
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: Please rename it.

File: iquip/apps/builder.py [hunk 33]
```diff
@@ -301,9 +302,17 @@ class _ScanEntry(_BaseEntry):
           "selected": The name of the selected scannable type.
           "NoScan", "RangeScan", "CenterScan", and "ExplicitScan": The dictionary that contains 
             argument info of the corresponding scannable type.
-        stackWidget: The QStackedWidget that contains widgets of each scannable type.
-        rangeWidget: TODO(AIJUH): The Widget that will be removed at next PR.
+        stackedWidget: The QStackedWidget that contains widgets of each scannable type.
+        scanButtonGroup: The QButtonGroup that groups QRadiobuttons for selecting scan widget.
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: "a scan widget" or "scan widgets".

**Review by BECATRUE — CHANGES_REQUESTED** [review]
Almost final!

File: iquip/apps/builder.py (hunk 30 above)
Comment (BECATRUE) [reply iquip/apps/builder.py]: If you want to use "at radiobutton group", "in the button group" looks better.
And, please write "is clicked".

File: iquip/apps/builder.py [hunk 34]
```diff
@@ -406,15 +408,15 @@ def value(self) -> Dict[str, Any]:
...
-        Once the checked button at radiobutton group clicked, this is called.
+        Once a scan type button at radiobutton group clicked, this is called.
 
         Attributes:
-            selectedButton: A QRadioButton that is clicked.
+            selectedButton: The clicked QRadioButton.
         """
-        selectedEnum = _ScanEntry.ScanType(self.scanButtonGroup.id(selectedButton))
-        self.stackedWidget.setCurrentWidget(self.scanWidgets[selectedEnum.name])
+        selectedType = _ScanEntry.ScanType(self.scanButtonGroup.id(selectedButton))
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: You renamed the same variable in `_ScanEntry.__init__()` to `selectedScanType`, hence please rename this to the same.

**Review by BECATRUE — CHANGES_REQUESTED** [review]
Final!

**Comment by Aijuh** [comment]
I've modified the code to return the selected arguments in the GUI when submitting an experiment.  The arguments below are submitted to the Artiq proxy.

image

File: iquip/apps/builder.py (hunk 30 above)
Comment (BECATRUE) [reply iquip/apps/builder.py]: @Aijuh

File: iquip/apps/builder.py [hunk 35]
```diff
@@ -403,8 +403,8 @@ def value(self) -> Dict[str, Any]:
         
         Returns the dictionary of the selected scannable arguments.
         """
-        selected = self.state["selected"]
-        return self.state[selected]
+        selectedWidget = self.stackedWidget.currentWidget()
```
Comment (BECATRUE) [inline iquip/apps/builder.py]: How about `selectedScanWidget` as you referred this as `scanWidget` in `__init__()`?

File: iquip/apps/builder.py [hunk 36]
```diff
@@ -564,10 +565,12 @@ def scanArguments(self) -> Dict[str, Any]:
         Returns the arguments of the range scan.
         """
         return {
+            "ty": "RangeScan",
             "start": self.startSpinBox.value(),
             "stop": self.stopSpinBox.value(),
             "npoints": self.npointsSpinBox.value(),
-            "randomize": self.randomizeCheckBox.isChecked()
+            "randomize": self.randomizeCheckBox.isChecked(),
+            "seed": None
```
Comment (BECATRUE) [inline iquip/apps/builder.py:573]: As I said before, the feature to set the seed manually looks necessary, hence please leave this as an issue.

**Review by BECATRUE — CHANGES_REQUESTED** [review]
I totally forgot `value()` should be updated. Thank you👍

**Comment by Aijuh** [comment]
I updated code and made issue about `seed` arguments.

**Review by BECATRUE — APPROVED** [review]
LGTM!
